README.md
# PointNet-Series

PointNet 系列的模型，以PointNet作为基线

## 更新

- 2022-01-20
  - 创建 pointnet_utils.py 文件
  - 创建 modelnet_prepare.py 文件
- 2022-01-19
  - 创建仓库：README.md, .gitignore, LICENSE 文件
  - 创建 modelnet_dataset.py 文件


HISTORY.md
# 项目日志

## 工作日志

- 2022-01-20
  - 创建 pointnet_utils.py 文件
    - 空间变换网络（Spatial Transformer Networks，STN）：STNkd 类
    - PointNet的编码器：PointNetEncoder 类
  - 创建 test_pointnet_utils.py 测试文件
    - test_stn3d()
    - test_stnkd()
    - test_PointNetEncoder_Classification()
    - test_PointNetEncoder_Segmentation()
  - 创建 modelnet_prepare.py 文件
- 2022-01-19
  - 创建仓库：README.md, .gitignore, LICENSE 文件
  - 创建 modelnet_dataset.py 文件
    - ModelNetDataset 类
  - 创建 test_modelnet_dataset.py 测试文件
    - test_modelnet10_test_1024pts()
    - test_modelnet10_test_1024pts_fps()
    - test_modelnet10_train_1024pts()
    - test_modelnet10_train_1024pts_fps()
    - test_modelnet40_test_1024pts()
    - test_modelnet40_test_1024pts_fps()
    - test_modelnet40_train_1024pts()
    - test_modelnet40_train_1024pts_fps()

model/test_pointnet_utils.py
"""
=================================================
@path   : PointNet-Series -> test_pointnet_utils.py
@IDE    : PyCharm
@Author : zYx.Tom, 526614962@qq.com
@Date   : 2022-01-19 15:13
@Version: v0.1
@License: (C)Copyright 2020-2022, zYx.Tom
@Reference:
@Desc   :
==================================================
"""
import unittest

import torch
from torch.autograd import Variable

from model.pointnet_utils import PointNetEncoder, STNkd, feature_transform_regularize
from tools import log_title, log_subtitle, log_info, log_debug
import logging

logging.basicConfig(
        level=logging.DEBUG,
        filename='../log/model/pointnet_utils.log',
        filemode='w',
        format="%(asctime)s - %(name)s - %(levelname)-9s - %(filename)-8s : %(lineno)s line - %(message)s",
        datefmt="%Y-%m-%d %H:%M:%S"
)


class MyTestCase(unittest.TestCase):
    def setUp(self) -> None:
        self.batch_size = 32
        self.num_points = 2500

    def log_method(self, title, sim_data, out):
        log_subtitle(title)
        log_debug(f"sim_data.size={sim_data.size()}")
        log_debug(f"stn.size={out.size()}")
        try:
            log_debug(f"loss={feature_transform_regularize(out)}")
        except Exception:
            log_debug(f"loss=null")
            pass

    def test_stn3d(self):
        # 批处理的大小：32条数据；3个特征；每条数据的数据点个数：2500个点
        title = "stn3d-data"
        channel = 3
        sim_data = Variable(torch.rand(self.batch_size, channel, self.num_points))
        trans = STNkd(channel)
        out = trans(sim_data)
        self.log_method(title, sim_data, out)
        self.assertEqual(out.detach().numpy().shape, (self.batch_size, channel, channel), "特征变换矩阵")

    def test_stnkd(self):
        # 批处理的大小：32条数据；64个特征；每条数据的数据点个数：2500个点
        title = "stnkd-data"
        channel = 64
        sim_data = Variable(torch.rand(self.batch_size, channel, self.num_points))
        trans = STNkd(channel)
        out = trans(sim_data)
        self.log_method(title, sim_data, out)
        self.assertEqual(out.detach().numpy().shape, (self.batch_size, channel, channel), "特征变换矩阵")

    def test_PointNetEncoder_Classification(self):
        # 批处理的大小：32条数据；3个特征；每条数据的数据点个数：2500个点
        title = "PointNetEncoder"
        in_channel, out_channel, trans_channel = 3, 1024, 64
        sim_data = Variable(torch.rand(self.batch_size, in_channel, self.num_points))

        net_encoder = PointNetEncoder(global_feat=True, feature_transform=True, channel=in_channel)
        out_encoder, trans_matrix, trans_feature = net_encoder(sim_data)
        self.log_method(title, sim_data, out_encoder)
        self.assertEqual(out_encoder.detach().numpy().shape,
                         (self.batch_size, out_channel),
                         "分类问题，增加第二个特征变换矩阵，全局特征")

        net_encoder = PointNetEncoder(global_feat=True, feature_transform=False, channel=in_channel)
        out_encoder, trans_matrix, trans_feature = net_encoder(sim_data)
        self.log_method(title, sim_data, out_encoder)
        self.assertEqual(out_encoder.detach().numpy().shape,
                         (self.batch_size, out_channel),
                         "分类问题，没有第二个特征变换矩阵，全局特征")

    def test_PointNetEncoder_Segmentation(self):
        # 批处理的大小：32条数据；3个特征；每条数据的数据点个数：2500个点
        title = "PointNetEncoder"
        in_channel, out_channel, trans_channel = 3, 1024, 64
        sim_data = Variable(torch.rand(self.batch_size, in_channel, self.num_points))

        net_encoder = PointNetEncoder(global_feat=False, feature_transform=True, channel=in_channel)
        out_encoder, trans_matrix, trans_feature = net_encoder(sim_data)
        self.log_method(title, sim_data, out_encoder)
        self.assertEqual(out_encoder.detach().numpy().shape,
                         (self.batch_size, out_channel + trans_channel, self.num_points),
                         "分割问题，增加第二个特征变换矩阵，全局特征+局部特征")

        net_encoder = PointNetEncoder(global_feat=False, feature_transform=False, channel=in_channel)
        out_encoder, trans_matrix, trans_feature = net_encoder(sim_data)
        self.log_method(title, sim_data, out_encoder)
        self.assertEqual(out_encoder.detach().numpy().shape,
                         (self.batch_size, out_channel + trans_channel, self.num_points),
                         "分割问题，没有第二个特征变换矩阵，全局特征")

    def test_something(self):
        self.assertEqual(True, False)  # add assertion here


if __name__ == '__main__':
    log_title("stnkd-model")
    global_feature = True
    unittest.main()


data_utils/test_modelnet_dataset.py
"""
=================================================
@path   : PointNet-Series -> test_modelnet_dataset.py
@IDE    : PyCharm
@Author : zYx.Tom, 526614962@qq.com
@Date   : 2022-01-19 15:13
@Version: v0.1
@License: (C)Copyright 2020-2022, zYx.Tom
@Reference:
@Desc   :
==================================================
"""
from argparse import Namespace
from datetime import datetime

from data_utils.modelnet_dataset import ModelNetDataset


class TestClass:
    def test_modelnet10_test_1024pts(self):
        args = Namespace()
        args.uniform_sample = False
        args.modelnet10 = True
        data_path = "../data/pickle/modelnet/"
        myDataset = ModelNetDataset(root=data_path, args=args, split='test')
        assert myDataset.get_pickle_file_name() == "../data/pickle/modelnet/modelnet10_test_1024pts.dat"

    def test_modelnet10_test_1024pts_fps(self):
        args = Namespace()
        args.uniform_sample = True
        args.modelnet10 = True
        data_path = "../data/pickle/modelnet/"
        myDataset = ModelNetDataset(root=data_path, args=args, split='test')
        assert myDataset.get_pickle_file_name() == "../data/pickle/modelnet/modelnet10_test_1024pts_fps.dat"

    def test_modelnet10_train_1024pts(self):
        args = Namespace()
        args.uniform_sample = False
        args.modelnet10 = True
        data_path = "../data/pickle/modelnet/"
        myDataset = ModelNetDataset(root=data_path, args=args, split='train')
        assert myDataset.get_pickle_file_name() == "../data/pickle/modelnet/modelnet10_train_1024pts.dat"

    def test_modelnet10_train_1024pts_fps(self):
        args = Namespace()
        args.uniform_sample = True
        args.modelnet10 = True
        data_path = "../data/pickle/modelnet/"
        myDataset = ModelNetDataset(root=data_path, args=args, split='train')
        assert myDataset.get_pickle_file_name() == "../data/pickle/modelnet/modelnet10_train_1024pts_fps.dat"

    def test_modelnet40_test_1024pts(self):
        args = Namespace()
        args.uniform_sample = False
        args.modelnet10 = False
        data_path = "../data/pickle/modelnet/"
        myDataset = ModelNetDataset(root=data_path, args=args, split='test')
        assert myDataset.get_pickle_file_name() == "../data/pickle/modelnet/modelnet40_test_1024pts.dat"

    def test_modelnet40_test_1024pts_fps(self):
        args = Namespace()
        args.uniform_sample = True
        args.modelnet10 = False
        data_path = "../data/pickle/modelnet/"
        myDataset = ModelNetDataset(root=data_path, args=args, split='test')
        assert myDataset.get_pickle_file_name() == "../data/pickle/modelnet/modelnet40_test_1024pts_fps.dat"

    def test_modelnet40_train_1024pts(self):
        args = Namespace()
        args.uniform_sample = False
        args.modelnet10 = False
        data_path = "../data/pickle/modelnet/"
        myDataset = ModelNetDataset(root=data_path, args=args, split='train')
        assert myDataset.get_pickle_file_name() == "../data/pickle/modelnet/modelnet40_train_1024pts.dat"

    def test_modelnet40_train_1024pts_fps(self):
        args = Namespace()
        args.uniform_sample = True
        args.modelnet10 = False
        data_path = "../data/pickle/modelnet/"
        myDataset = ModelNetDataset(root=data_path, args=args, split='train')
        assert myDataset.get_pickle_file_name() == "../data/pickle/modelnet/modelnet40_train_1024pts_fps.dat"


def main(name):
    print(f'Hi, {name}', datetime.now())

    pass


if __name__ == "__main__":
    __author__ = 'zYx.Tom'
    main(__author__)


tools.py
"""
=================================================
@path   : PointNet-Series -> tools.py
@IDE    : PyCharm
@Author : zYx.Tom, 526614962@qq.com
@Date   : 2022-01-20 11:09
@Version: v0.1
@License: (C)Copyright 2020-2022, zYx.Tom
@Reference:
@Desc   :
==================================================
"""
import logging
from datetime import datetime

# 默认的warning级别，只输出warning以上的
# 使用basicConfig()来指定日志级别和相关信息
logging.basicConfig(
        level=logging.DEBUG,
        filename='../log/demo.log',
        filemode='w',
        format="%(asctime)s - %(name)s - %(levelname)-9s - %(filename)-8s : %(lineno)s line - %(message)s",
        datefmt="%Y-%m-%d %H:%M:%S"
)


def show_subtitle(message):
    # 输出运行模块的子标题
    print('-' * 15, '>' + message + '<', '-' * 15)
    pass


def show_title(message):
    # 输出运行模块的子标题
    print()
    print('=' * 15, '>' + message + '<', '=' * 15)
    pass


def log_title(message):
    message = '=' * 15 + '>' + message + '<' + '=' * 15
    logging.info(message)


def log_subtitle(message):
    message = '-' * 15 + '>' + message + '<' + '-' * 15
    logging.info(message)


def log_debug(message):
    logging.debug(message)


def log_info(message):
    logging.info(message)


def beep_end():
    # 运行结束的提醒
    import winsound
    winsound.Beep(600, 500)
    pass


def main(name):
    print(f'Hi, {name}', datetime.now())
    show_title("title")
    show_subtitle("subtitle")
    pass


if __name__ == "__main__":
    __author__ = 'zYx.Tom'
    main(__author__)


model/pointnet_utils.py
# -*- encoding: utf-8 -*-
"""
=================================================
@path   : PointNet-Series -> pointnet_utils.py
@IDE    : PyCharm
@Author : zYx.Tom, 526614962@qq.com
@Date   : 2022-01-20 11:02
@Version: v0.1
@License: (C)Copyright 2020-2022, zYx.Tom
@Reference:
@Desc   :
==================================================
"""
from datetime import datetime
import torch
import torch.nn as nn
import torch.nn.functional as F

from torch.autograd import Variable


class STNkd(nn.Module):
    """
    空间变换网络（Spatial Transformer Networks，STN）
    """

    def __init__(self, channel=3, k=3):
        """
        :param channel: 输入数据的通道数，特征变换的维度
        """
        super(STNkd, self).__init__()

        self.channel = channel

        self.conv1 = torch.nn.Conv1d(channel, 64, (1,))
        self.conv2 = torch.nn.Conv1d(64, 128, (1,))
        self.conv3 = torch.nn.Conv1d(128, 1024, (1,))

        self.fc1 = nn.Linear(1024, 512)
        self.fc2 = nn.Linear(512, 256)
        self.fc3 = nn.Linear(256, channel * channel)

        self.relu = nn.ReLU()

        self.bn1 = nn.BatchNorm1d(64)
        self.bn2 = nn.BatchNorm1d(128)
        self.bn3 = nn.BatchNorm1d(1024)
        self.bn4 = nn.BatchNorm1d(512)
        self.bn5 = nn.BatchNorm1d(256)
        pass

    def forward(self, x):
        """
        :return: 空间变换矩阵
        """
        batch_size = x.size()[0]
        x = F.relu(self.bn1(self.conv1(x)))
        x = F.relu(self.bn2(self.conv2(x)))
        x = F.relu(self.bn3(self.conv3(x)))
        x = torch.max(x, 2, keepdim=True)[0]
        x = x.view(-1, 1024)

        x = F.relu(self.bn4(self.fc1(x)))
        x = F.relu(self.bn5(self.fc2(x)))
        x = self.fc3(x)

        # 单位矩阵 展开成 序列
        identify_list = Variable(torch.eye(self.channel).view(1, -1).repeat(batch_size, 1))
        identify_list = identify_list.cuda() if x.is_cuda else identify_list
        x = x + identify_list
        x = x.view(-1, self.channel, self.channel)
        return x


class PointNetEncoder(nn.Module):
    """
    PointNet的编码器
    """

    def __init__(self, global_feat=True, feature_transform=True, channel=3):
        """

        :param global_feat: 分类问题，只生成全局特征；分割问题，拼接生成的局部特征+全局特征
        :param feature_transform: 第二个变换矩阵，特征变换矩阵
        :param channel: 输入数据的通道数
        """
        super(PointNetEncoder, self).__init__()
        self.stn = STNkd(channel=3)
        self.conv1 = torch.nn.Conv1d(channel, 64, 1)
        self.conv2 = torch.nn.Conv1d(64, 128, 1)
        self.conv3 = torch.nn.Conv1d(128, 1024, 1)
        self.bn1 = nn.BatchNorm1d(64)
        self.bn2 = nn.BatchNorm1d(128)
        self.bn3 = nn.BatchNorm1d(1024)
        self.global_feature = global_feat
        self.feature_transform = feature_transform
        if feature_transform:
            self.fstn = STNkd(channel=64)

    def forward(self, x):
        B, D, N = x.size()
        # 第一个变换矩阵，空间对齐矩阵
        trans_matrix = self.stn(x)
        x = x.transpose(2, 1)

        if D > 3:
            feature = x[:, :, 3:]
            x = x[:, :, :3]
            x = torch.bmm(x, trans_matrix)
            x = torch.cat([x, feature], dim=2)
        else:
            x = torch.bmm(x, trans_matrix)

        x = x.transpose(2, 1)
        x = F.relu(self.bn1(self.conv1(x)))

        # 第二个变换矩阵，特征变换矩阵
        if self.feature_transform:
            trans_feature = self.fstn(x)
            x = x.transpose(2, 1)
            x = torch.bmm(x, trans_feature)
            x = x.transpose(2, 1)
        else:
            trans_feature = None

        point_feature = x
        x = F.relu(self.bn2(self.conv2(x)))

        # 全连接网络+bn
        x = self.bn3(self.conv3(x))
        # 全局池化层
        x = torch.max(x, 2, keepdim=True)[0]
        x = x.view(-1, 1024)
        if self.global_feature:
            return x, trans_matrix, trans_feature
        else:
            x = x.view(-1, 1024, 1).repeat(1, 1, N)
            return torch.cat([x, point_feature], 1), trans_matrix, trans_feature


def feature_transform_regularize(trans):
    #
    d = trans.size()[1]
    I = torch.eye(d)[None, :, :]
    I = I.cuda() if trans.is_cuda else I
    loss = torch.mean(torch.norm(torch.bmm(trans, trans.transpose(2, 1)) - I, dim=(1, 2)))
    return loss


def main(name):
    print(f'Hi, {name}', datetime.now())
    pass


if __name__ == "__main__":
    __author__ = 'zYx.Tom'
    main(__author__)


data_utils/modelnet_prepare.py
"""
=================================================
@path   : PointNet-Series -> modelnet_prepare.py
@IDE    : PyCharm
@Author : zYx.Tom, 526614962@qq.com
@Date   : 2022-01-19 15:54
@Version: v0.1
@License: (C)Copyright 2020-2022, zYx.Tom
@Reference:
https://shapenet.cs.stanford.edu/media/modelnet40_normal_resampled.zip
@Desc   : 预处理数据，加工多个以文件保存的数据为单个列表保存的数据
==================================================
"""
import pickle
import numpy as np
import argparse
from datetime import datetime
from os.path import join

from tqdm import tqdm


def farthest_point_sample(xyz, npoint):
    """
    最远点采样(N:Number, D:Dimension)
    不断迭代地选择距离已有采样点集合的最远点，从而得到物体的轮廓
    采样点的数量应该小于原始点的数量，否则就是取出所有的点
    Input:
        xyz: point cloud data, [N, D]
        npoint: number of samples
    Return:
        centroids: sampled point cloud index, [npoint, D]
    """
    N, D = xyz.shape
    xyz = xyz[:, :3]
    centroids = np.zeros((npoint,))  # 质心点
    distance = np.ones((N,)) * 1e10  # 采样点到所有点的距离最大值
    # farthest = np.random.randint(0, N)  # 随机初始化一个点作为首个质心点

    barycenter = (np.sum(xyz, 0) / xyz.shape[0])  # 计算形状的质心坐标 及 距离质心最远的点
    dist = np.sum((xyz - barycenter) ** 2, -1)  # 计算所有点到形状质心的距离
    farthest = np.argmax(dist, -1)  # 将距离质心最远的点作为第一个点（比随机初始化更加精确）

    # 每一次循环找的是当前质心点的最远点
    # distance里存的是每一次操作以后N个点相对于所有备选点集的最远距离
    # mask存在是因为点到点集的距离是点到点集中每个点的距离中的最小值
    for i in range(npoint):
        centroids[i] = farthest  # 更新第 i 个最远点
        centroid = xyz[farthest, :]  # 取出这个最远点的xyz坐标
        dist = np.sum((xyz - centroid) ** 2, -1)  # 计算点集中所有的点到这个质心点的距离
        mask = dist < distance
        distance[mask] = dist[mask]  # 更新最远点与整个点集之间的最短距离
        farthest = np.argmax(distance, -1)  # 返回最远点的索引
    xyz = xyz[centroids.astype(np.int32)]  # 返回坐标
    return xyz


def main(name):
    print(f'Hi, {name}', datetime.now())
    prepare_data()
    # origin_xyz = np.random.randn(5, 3)
    # print(origin_xyz)
    # print(farthest_point_sample(origin_xyz, 3))


def prepare_data():
    args = argparse.Namespace()
    print("将 ModelNet 数据加工成一个数组，方便存储和使用")
    root = "../data/modelnet40_normal_resampled"
    num_points = 1024
    modelnet_cat = "modelnet10"
    process_data(root, num_points, modelnet_cat, split='train', fps='')
    process_data(root, num_points, modelnet_cat, split='test', fps='')
    process_data(root, num_points, modelnet_cat, split='train', fps='_fps')
    process_data(root, num_points, modelnet_cat, split='test', fps='_fps')
    modelnet_cat = "modelnet40"
    process_data(root, num_points, modelnet_cat, split='train', fps='')
    process_data(root, num_points, modelnet_cat, split='test', fps='')
    process_data(root, num_points, modelnet_cat, split='train', fps='_fps')
    process_data(root, num_points, modelnet_cat, split='test', fps='_fps')


def process_data(root, num_points, modelnet_cat, split, fps):
    """
    将modelnet中的数据加工到列表变量中，方便存储与使用
    :param root: 数据所在目录
    :param num_points:
    :param modelnet_cat: 数据是10个类别，还是40个类别
    :param split: 数据用于训练（train）还是测试（test）
    :param fps: 使用最远点采样（farthest point sample）
    :return: 处理成功或者失败
    """
    # 从 modelnet10_shape_names.txt，modelnet40_shape_names.txt 取出形状类别
    shape_category = [line.rstrip() for line in open(join(root, f"{modelnet_cat}_shape_names.txt"))]
    # 对形状类别进行编号
    shape_classes = dict(zip(shape_category, range(len(shape_category))))
    # 从 modelnet10_train.txt, modelnet10_test.txt, modelnet40_train.txt, modelnet40_test.txt 中取出形状列表
    shape_ids = [line.rstrip() for line in open(join(root, f"{modelnet_cat}_{split}.txt"))]
    # 将列表中每个形状的类别提出
    shape_names = ['_'.join(file_name.split('_')[0:-1]) for file_name in shape_ids]
    # 生成每个形状文件的访问路径
    data_path = [(shape_names[i], join(root, shape_names[i], shape_ids[i] + '.txt'))
                 for i in range(len(shape_ids))]
    # 保存 pickle 文件的路径
    save_path_str = f"../data/tmp/{modelnet_cat}_{split}_{num_points}pts{fps}.dat"
    # 初始化保存所有数据（点、标签）的列表变量
    num_data = len(shape_ids)
    list_of_points = [None] * num_data
    list_of_labels = [None] * num_data
    # 读入所有数据
    for index in tqdm(range(num_data), total=num_data):
        file_name = data_path[index]
        shape_class = np.array([shape_classes[data_path[index][0]]]).astype(np.int32)
        point_set = np.loadtxt(file_name[1], delimiter=',').astype(np.float32)
        # 按默认顺序采样，还是按最远点采样
        list_of_points[index] = point_set[0:num_points, :] \
            if fps == "" else farthest_point_sample(point_set, num_points)
        list_of_labels = shape_class
        pass
    # 写入文件中
    with open(save_path_str, 'wb') as f:
        pickle.dump([list_of_points, list_of_labels], f)


if __name__ == "__main__":
    __author__ = 'zYx.Tom'
    main(__author__)


data_utils/modelnet_dataset.py
"""
=================================================
@path   : PointNet-Series -> modelnet_dataset.py
@IDE    : PyCharm
@Author : zYx.Tom, 526614962@qq.com
@Date   : 2022-01-19 10:26
@Version: v0.1
@License: (C)Copyright 2020-2022, zYx.Tom
@Reference:
@Desc   :
==================================================
"""
import argparse
import logging
from datetime import datetime
import torch.utils.data as data
from os.path import join
import pickle
from torch.utils.data import DataLoader


class ModelNetDataset(data.Dataset):
    def __init__(self, root, args, split='train'):
        prefix = "modelnet10" if args.modelnet10 else "modelnet40"
        fps = "_fps" if args.uniform_sample else ""
        self.pickle_file_name = join(root, f'{prefix}_{split}_1024pts{fps}.dat')
        try:
            with open(self.pickle_file_name, 'rb') as f:
                self.list_of_points, self.list_of_labels = pickle.load(f)
        except FileNotFoundError:
            print("请先完成数据预处理！")

    def __len__(self):
        return len(self.list_of_points)

    def __getitem__(self, item):
        return self.list_of_points[item], self.list_of_labels[item]

    def get_pickle_file_name(self):
        return self.pickle_file_name


def main(name):
    """测试ModelNetDataset类"""
    print(f'Hi, {name}', datetime.now())

    """参数"""
    args = argparse.Namespace()
    args.uniform_sample = False
    args.modelnet10 = True

    def log_string(info):
        logger.info(info)
        print(info)
        pass

    """日志"""
    logger = logging.getLogger("ModelNet")
    logger.setLevel(logging.INFO)
    formatter = logging.Formatter("%(asctime)s - %(name)s -%(levelname)s - %(message)s")
    file_handler = logging.FileHandler("../log/data/ModelNet/eval.txt")
    file_handler.setLevel(logging.INFO)
    file_handler.setFormatter(formatter)
    logger.addHandler(file_handler)
    log_string("参数。。。")
    log_string(args)

    """数据加载"""
    log_string("载入数据集。。。")
    data_path = "../data/pickle/modelnet/"
    myDataset = ModelNetDataset(root=data_path, args=args, split='test')
    print(myDataset.get_pickle_file_name())
    data_iter = DataLoader(myDataset, batch_size=15, shuffle=True)
    for i_batch, batch_data in enumerate(data_iter):
        print(i_batch)
        print(batch_data[0])
        print(batch_data[1])
        if i_batch > 2:
            break
    pass


if __name__ == "__main__":
    __author__ = 'zYx.Tom'
    main(__author__)
